core-backend/src/funcs.rs

```diff
@@ -616,55 +574,60 @@ where
         )
     }
 
-    pub fn read(at: u32, len: u32, buffer_ptr: u32) -> impl Syscall<Caller> {
-        FallibleSyscall::new::<ErrorBytes>(CostToken::Read, move |ctx: &mut CallerWrap<Caller>| {
-            let payload_lock = ctx.ext_mut().lock_payload(at, len)?;
-            payload_lock
-                .drop_with::<MemoryAccessError, _>(|payload_access| {
-                    let mut f = || {
-                        let mut registry = MemoryAccessRegistry::default();
-                        let write_buffer = registry.register_write(buffer_ptr, len);
-                        let mut io = registry.pre_process(ctx)?;
-                        io.write(ctx, write_buffer, payload_access.as_slice())?;
-                        Ok(())
-                    };
-                    let res = f();
-                    let unlock_bound = ctx.ext_mut().unlock_payload(payload_access.into_lock());
-
-                    DropPayloadLockBound::from((unlock_bound, res))
-                })
-                .into_inner()
-                .map_err(Into::into)
-        })
+    pub fn read(at: u32, buffer: WriteInGrRead) -> impl Syscall<Caller> {
+        FallibleSyscall::new::<ErrorBytes>(
+            CostToken::Read,
+            move |ctx: &mut MemoryCallerContext<Caller>| {
+                let payload_lock = ctx.caller_wrap.ext_mut().lock_payload(at, buffer.size())?;
+                payload_lock
+                    .drop_with::<MemoryAccessError, _>(move |payload_access| {
+                        let f = || {
+                            buffer.write(ctx, payload_access.as_slice())?;
+                            Ok(())
+                        };
+                        let res = f();
+                        let unlock_bound = ctx
+                            .caller_wrap
+                            .ext_mut()
+                            .unlock_payload(payload_access.into_lock());
+
+                        DropPayloadLockBound::from((unlock_bound, res))
+                    })
+                    .into_inner()
+                    .map_err(Into::into)
+            },
+        )
     }
 
-    pub fn size(size_ptr: u32) -> impl Syscall<Caller> {
-        InfallibleSyscall::new(CostToken::Size, move |ctx: &mut CallerWrap<Caller>| {
-            let size = ctx.ext_mut().size()? as u32;
+    pub fn size(size_write: WriteAs<[u8; 4]>) -> impl Syscall<Caller> {
```

Comment:
WriteAS< u32 > doesnt work? u32: Encode resulting in le bytes

Reply:
Yeah, you're right, converting to bytes was redundant here